### ISSUE TITLE
Optimize stloc, ldloc to dup, stloc

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#21764]
This only happens when you use inline assignments, but the compiler could do that even if it's not an inline assignment.

**Comment by mikedn** [comment]
What exactly is this optimization supposed to improve? I'm not sure it would be beneficial as far as the JIT is concerned, the use of `dup` may result in the creation of an additional local var.

**Comment by sharwell** [comment]
:memo: Before implementing this change, we should first investigate the JIT output. If the JIT handles this IL sequence by emitting a store without actually reloading from the local variable, then there would be no reason to change it.

**Comment by alrz** [comment]
I found this blog post for a micro-benchmark on that. 

https://www.danielcrabtree.com/blog/164/c-sharp-7-micro-benchmarking-the-three-ways-to-cast-safely

**Comment by mikedn** [comment]
Doing benchmarks and then analyzing the IL to draw conclusions is complete nonsense.

**Comment by sharwell** [comment]
@alrz Does the "as" approach still use stloc/ldloc when you declare 4 local variables before it? stloc.0 and dup are both one-byte instructions, but stloc.s is longer.

**Comment by alrz** [comment]
@sharwell 

>Does the "as" approach still use stloc/ldloc when you declare 4 local variables before it?

Yes. That's not the only case though, when you name an expression by a variable and `if` on that afterwards for example, you'd still get stloc/ldloc. I think this belogs to broader optimizations for "inlining" expressions under some circumstances.


@mikedn 

>Doing benchmarks and then analyzing the IL to draw conclusions is complete nonsense.

I wish you'd elaborate. I think generated IL does give you an idea about possible improvements.

>the use of dup may result in the creation of an additional local var.

I don't follow. AFAIK dup uses stack, not locals.

**Comment by mikedn** [comment]
> I don't follow. AFAIK dup uses stack, not locals.

The JIT compiler (and pretty much all other native code generators/optimizers) does not use a stack based representation. Once the IL is imported into the internal compiler representation there is no `dup`, `pop` or anything else that relates to the abstract IL stack.

As such, `dup` is transformed into the equivalent of a `stloc.x`/`ldloc.x` pair where `x` is a temporary variable created by the JIT. AFAIR there is one exception - if the value that's on the stack happens to be produced by a `ldloc.n` (where `n` is an existing local variable) then `dup` is transformed into another `ldloc.n`

The added temporary variable may or may not have an impact on the generated code. It's somewhat difficult to evaluate this as the impact may only be visible in larger methods. In general, the more local variable a method contains the more likely is to run into optimization issues because the JIT tracks only a limited number of locals in its analysis of the code.

>  I think generated IL does give you an idea about possible improvements.

Without knowing how the JIT behaves it is pretty much impossible to say if a change to the IL code is an improvement or not. For example, the JIT may recognize and optimize certain IL patterns that it knows that managed compilers tend to emit. Or it may have internal limits such as the number of locals it tracks or the maximum depth of an expression tree it allows.

And to be clear, I'm not saying that using `dup` is necessarily a bad thing. The main problem is proving that it is not.

**Comment by sharwell** [comment]
> EDIT: this could affect atomic property of the code, i.e. dup, stloc would always operate on the exact same value, but stloc, ldloc may not.

I'm having trouble finding ways where this could produce observably incorrect behavior. For example, the following code: 

```csharp
string text;
if ((text = input as string) != null)
{
  return text.Length;
}
```

Actually results in the use of the dup/stloc sequence.

At this point, the one thing that would definitely go in the "pro" column for dup/stloc is the bytecode sequence is more succinct in some cases. However, there is a big drawback as well. One thing you'll observe in the "As" approach is the bytecode sequence being discussed actually spans *two* statements in the C# source code:

```
# string text = input as string;
IL_0001: isinst [mscorlib]System.String
IL_0006: stloc.0

# if (text != null)
IL_0007: ldloc.0
...
```

The optimization¹ which would result in the use of dup/stloc is exactly the one you mentioned: ability to inline expressions during IL generation. In particular, this would be part of handling cases where a statement assigns a value to a local variable, and then that variable is immediately read.

My overall observation on this issue is currently the following:

* The current generated code sequences make sense for a direct and correct conversion of source code to IL.
* The bytecode representation for the current generated code sequence is larger than necessary for some code paths in some methods, but for many cases is not. The biggest impact would affect methods where the use of `as` followed by a null check precludes the use of the "Tiny Format" for method metadata (ECMA-335 §II.25.4.2) due to failure to eliminate a local variable in Release build configurations.
* The new pattern-matching `as` expression appears to cover the majority of cases where users would want this optimization to apply.

I would conclude from this that the best course of action is to leave the code generation for "as" followed by a null check unchanged for now, and instead continue working to improve the generated code for pattern matching expressions in edge cases that are known to produce suboptimal code.

¹ I'm making no claims regarding the performance impact of the optimization here.

**Comment by alrz** [comment]
>I'm having trouble finding ways where this could produce observably incorrect behavior. 

I think so because there are "locals" after all. however, if we want to do this for stfld/ldfld then I think the issue arise.

>I would conclude from this that the best course of action is to leave the code generation for "as" followed by a null check unchanged for now

Note that this is not just as/null-check, when you assign a value to a local, and then immediately read from it this happens.
